test/common/TestCommonEqEpsilon.hs

```diff
@@ -35,6 +36,15 @@ eqEpsilonRef = unsafePerformIO $ newIORef eqEpsilonDefault
 setEpsilonEq :: EqEpsilon -> IO ()
 setEpsilonEq (EqEpsilon x) = atomicWriteIORef eqEpsilonRef x
 
+exceptionHandler :: SomeException -> IO ()
+exceptionHandler (SomeException e) = do
+  case e of
+    -- The wildcard branch below prints:
+    -- Hello from exceptionHandler! Exception is: HUnitFailure (Just (SrcLoc {srcLocPackage = "horde-ad-0.1.0.0-inplace-testLibrary", srcLocModule = "TestCommonEqEpsilon", srcLocFile = "test/common/TestCommonEqEpsilon.hs", srcLocStartLine = 118, srcLocStartCol = 5, srcLocEndLine = 118, srcLocEndCol = 16})) (Reason "expected: 0.8991\n but got: 0.7424999999999999\n (maximum margin of error: 1.0e-6)") of type: HUnitFailure
+    -- BUT we cannot match e with HUnitFailure x y because it doesn't compile:
+    -- HUnitFailure x y -> putStrLn $ "We got HUnitFailure!"
```

Comment:
The compile error is:

```
test/common/TestCommonEqEpsilon.hs:49:5: error:
    • Couldn't match expected type ‘e’ with actual type ‘HUnitFailure’
      ‘e’ is a rigid type variable bound by
        a pattern with constructor:
          SomeException :: forall e. Exception e => e -> SomeException,
        in an equation for ‘exceptionHandler’
        at test/common/TestCommonEqEpsilon.hs:44:19-33
    • In the pattern: HUnitFailure x y
      In a case alternative:
          HUnitFailure x y -> putStrLn $ "We got HUnitFailure!"
      In a stmt of a 'do' block:
        case e of
          HUnitFailure x y -> putStrLn $ "We got HUnitFailure!"
          _ -> putStrLn
                 $ "Hello from exceptionHandler! Exception is: "
                     ++ (show e) ++ " of type: " ++ (show $ typeOf e)
    • Relevant bindings include
        e :: e (bound at test/common/TestCommonEqEpsilon.hs:44:33)
   |
49 |     HUnitFailure x y -> putStrLn $ "We got HUnitFailure!"
   |
```

and I thought: this is perhaps due to a missing instance, i.e. `instance Exception HUnitFailure`. So I tried:

`import           Test.Tasty.HUnit.Orig ()`

(where that instance is declared) but then I got:

```
test/common/TestCommonEqEpsilon.hs:15:1: error:
    Could not load module ‘Test.Tasty.HUnit.Orig’
    it is a hidden module in the package ‘tasty-hunit-0.10.0.3’
    it is a hidden module in the package ‘tasty-hunit-0.10.0.3’
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
   |
15 | import           Test.Tasty.HUnit.Orig ()
   |
```

and if I just declare the instance: `instance Exception HUnitFailure` then I get:

```
test/common/TestCommonEqEpsilon.hs:19:10: error:
    Duplicate instance declarations:
      instance Exception HUnitFailure
        -- Defined at test/common/TestCommonEqEpsilon.hs:19:10
      instance [safe] Exception HUnitFailure
        -- Defined in ‘tasty-hunit-0.10.0.3:Test.Tasty.HUnit.Orig’
   |
19 | instance Exception HUnitFailure
   |
```

Is there a way to pattern match on `SomeException (HUnitFailure)` then?...

What is `[safe]`?

BTW I just discovered [here](https://hackage.haskell.org/package/tasty-hunit-0.10.0.3/docs/src/Test.Tasty.HUnit.Orig.html):

```
-- | This is the code copied from the original hunit package (v. 1.2.5.2).
-- with minor modifications
```

but this is probably irrelevant.

Reply:
Is this now explained by there being two copies of HUnitFailure?

Reply:
No.

Reply:
`SomeException` fails because, with this quantification in place, you need to choose a type `e`, which decides at what point of the exception hierarchy you want to be. You choose with https://downloads.haskell.org/ghc/latest/docs/libraries/base-4.17.0.0/Control-Exception.html#v:fromException (or rather, with the type that the context implies in the application of `fromException`). In response you get an exception if, indeed, at this point of the hierarchy there is any exception encapsulated inside that expression. I think, If you implied another type `e`, you could get a value of another type out (though internally it would be the same untyped value, but you'd get a different API for it).

> it is a hidden module

Right, an internal hidden module.

> What is [safe]?

No idea.

> Is there a way to pattern match on SomeException (HUnitFailure) then?

I don't think so.

Reply:
I just managed to fix the whole thing without touching any exceptions at all.

Reply:
> `SomeException` fails because, with this quantification in place, you need to choose a type `e`, which decides at what point of the exception hierarchy you want to be. You choose with https://downloads.haskell.org/ghc/latest/docs/libraries/base-4.17.0.0/Control-Exception.html#v:fromException (or rather, with the type that the context implies in the application of `fromException`). In response you get an exception if, indeed, at this point of the hierarchy there is any exception encapsulated inside that expression. I think, If you implied another type `e`, you could get a value of another type out (though internally it would be the same untyped value, but you'd get a different API for it).

What do you say? That you cannot pattern match on `SomeException` (except the wildcard `_` pattern), but you have to explicitly call `fromException` instead? Why?

Reply:
> Why?

Because it's 21st century and we have OOP or at least hierarchies or at least subtyping or at least the type tricks above that emulate subtyping. But seriously, is the way hierarchy is expressed surprising or the purpose unclear or what else?

Reply:
Perhaps the crucial bit is **extensible** hierarchies. You have a virtual datatype of all possible exceptions you can get in your application and any new dependency you link against can add a few cases to that "datatype". That's can't be done with normal datatypes. OCaml extensible variant types might suffice, but Haskell's anonymous sums are probably not there yet.